autogen/main.tf.tmpl

```diff
@@ -44,8 +46,36 @@ resource "google_compute_global_address" "default" {
   count      = var.create_address ? 1 : 0
   project    = var.project
   name       = "${var.name}-address"
-  ip_version = var.ip_version
+  ip_version = "IPV4"
 }
+### IPv4 block ###
+
+### IPv6 block ###
+resource "google_compute_global_forwarding_rule" "http_ipv6" {
```

Comment:
We probably need to make all the ipv6 resources conditional so they are only created if:
1. ipv6_address is provided
2. `create_ipv6_address` is true.

Reply:
Makes sense - fixed.
* Added a new local boolean var `create_ipv6_resources` that is evaluated to true with
`((local.ipv6_address != nul)) && (length(local.ipv6_address)) > 0)`

`create_ipv6_resources` is then used in the fwd-rule resources to set the count to 0 or 1.

Reply:
Note that there are potential issues with dynamic dependencies if we do it this way (if users try to pass in a reference to their own resource). We might want to simply expose `use_ipv6` and `use_ipv4` variables.

Reply:
Well, when I tried running it last on GCP, the reference (self-link) to the newly created IPv4/IPv6 address didn't work - it required **explicit** IP address value.

This is also reflected in the corresponding docs to the parameters that you're expected to pass the actual IP address value.

Reply:
Maybe, we can think of more checks on the value? (for example, regex based string check)

Reply:
Right, but even referencing the value is potentially an issue (ex. `ipv6_address = google_compute_global_address.mine.address`) as it will raise the error of count referencing a dynamic value.

Reply:
Couple of use-cases I can think of where this configurability **may be** important:
1. bring-your-own-address - google supports it.
2. security considerations - where you might want to add firewall rules on-prem based on source/destination IP addresses - which might be difficult to do if this IP address keeps changing. So, a separately statically allocated elastic IP that's reviewed and approved by the security team helps here.

Let's fix this if we get any reports from users?

Reply:
Right, I don't disagree with the use case. I'm just concerned that having a conditional based on the value is likely to raise errors and will require a backwards-incompatible release.

Reply:
One way to look at it could be to treat IPv6 as equally essential as IPv4 - IPv4 address space is almost exhausted anyways and a lot of the mobile networks are running on IPv6 only - with IPv4 access provided via carrier-grade NATs.

If we take this approach, we can remove the conditional and force the IPv6 resources to be created/destroyed as first-class citizen instead of optional ones.

Reply:
IPv6 isn't free so I don't want to force everyone to use it.

Reply:
We already have a lot of variables to confuse the user:
1. ipv6_address - for custom address
2. create_ipv6_address - for creating new IPv6 address, if user doesn't have one already.
3. enable_ipv6 - the new one you proposed.

There is some interplay between these variables - and users are expected to keep track of them correctly.

I would prefer to actually cut down the variables controlling the same feature.

How about if we use:
1. `ipv6_address = "disable"` (default) to indicate no ipv6 configs to be deployed.
2. `ipv6_address = "new"` to indicate **creation** of the address
3. any other value as pre-created IPv6 address?

This could simplify the config the user is expected to fiddle with and we can then create the required local TF state variables based on that?

just a thought.

Reply:
Sure, that makes sense to me.

Reply:
Alright, let me rework the TF code with reduced no. of variables.

Reply:
@morgante I have refactored some of the variables and conditionals in commit 4f2fb4f593e9495ddfa53b9aeea8314dbf9b904d.

Please have a look.

Essentially, I have kept only 2 user-facing variables - as per our discussion above:

1. boolean flag `enable_ipv6` to control all of the IPv6 resources
2. string var `ipv6_address` with a sentinel value `"new"` to indicate allocation of new IPv6 EIP.

Reply:
Thanks. The major downside I see is that now we're referencing `ipv6_address` in a `count` conditional which means it *cannot* be a dynamic value (ex. `google_compute_global_address.mine.address`) which actually seems like a meaningful problem as I'd imagine it's a common scenario to manage the original IP address creation in Terraform.

The alternative would be to control creation with a separate `create_ipv6` variable.

That being said, i agree your interface is simpler. What do others think? @naseemkullah @bharathkkb ?

Reply:
I'm afraid I don't follow, what's the tl;dr as to why we cannot have `create_ipv6_address` and `ipv6_address`  to either create or use an existing ipv6 address? they would be mutually exclusive, and if both were left unset, ipv6 would be disabled.

Reply:
The issue is if we want to accept `ipv6_address` as an input variable we cannot make anything conditional on it.

Reply:
I don't fully get it tbh, but it seems there is no working around the unfortunate third value of `enable_ipv6` if we want to be able to dynamically reference an existing `ipv6_address`, if that's true I am in favour of it since being bale to dynamically reference is worth having to deal with the confusing vars.

Reply:
as it is, other than the **validation** of appropriate ipv6 address, I don't think we need to make the resources conditional on the variable `ipv6_address`. Do you think if we just go with `enable_ipv6` for all the conditionals would make it better/easier?

terraform doesn't seem to have a built-in ability similar to an `assert` or to raise an error and abort if validations fail - there are workarounds discussed here https://github.com/hashicorp/terraform/issues/15469.

@morgante @naseemkullah please let me know if you guys feel comfortable with only using `enable_ipv6` for conditionals and think about validating the ipv6 values later.
I'll adjust the tf code accordingly.

Reply:
If tf apply with an invalid ipv6 will fail on GCP's side, that is good enough IMO

Reply:
I have reworked the `count` conditionals to not depend on the `ipv6_address` in commit 8ed7b020f50612c3bcfd6718dcc4b16a122b6e22.

@morgante @naseemkullah please have a look and let me know if this works for you.

Reply:
Unfortunately the `google_compute_global_address` resource itself (line 86) is still conditional on the `ipv6_address` variable. It might be good to verify this yourself by updating one of the examples/tests to create an address outside the module then attempt to pass it into the module.

Reply:
@morgante I have added some examples here copied from the `https-redirect` example from this repo:
https://github.com/kunal-g/google-lb-http-ipv6-examples

TL;DR:
both situations seem to be working as expected:
* predefined IPv6 address (created using `gcloud compute address create` **outside** of this module) 
* newly created IPv6 address (using the `google_compute_global_address` resource reference **within** this module).

Please let me know if there are more changes needed. Happy to help.

Reply:
The situation I'm concerned about would be using the `google_compute_global_address` resource *outside* the module. Example:

```
resource "google_compute_global_address" "main" {
  project    = var.project
  name       = "my-outside-address"
  ip_version = "IPV6"
}

module "gce-lb-http" {
  source               = "github.com/kunal-g/terraform-google-lb-http?ref=issue-112-dual-stack"
  name                 = "ci-https-redirect"
  ipv6_address    = google_compute_global_address.main.address
  ...
}
```

Reply:
Yup, you are right on target - tried what you suggested above and got this error:
```
➜  https-redirect git:(main) ✗ terraform plan -var-file=test.tfvars
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.template_file.group-startup-script: Refreshing state...
module.mig.data.google_compute_zones.available: Refreshing state...
google_compute_network.default: Refreshing state... [id=projects/frontier-268910/global/networks/default]
module.mig_template.data.google_compute_image.image_family: Refreshing state...
module.mig_template.data.google_compute_image.image: Refreshing state...
google_compute_subnetwork.default: Refreshing state... [id=projects/frontier-268910/regions/asia-south1/subnetworks/default]

Error: Invalid count argument

  on .terraform/modules/gce-lb-http/main.tf line 82, in resource "google_compute_global_address" "default_ipv6":
  82:   count      = (local.create_ipv6_address) ? 1 : 0

The "count" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the count depends on.
```

Let me figure a way out. Thanks for highlighting the case.

Reply:
But, wouldn't the same problem occur for the `ipv4` address resources?
When I wrote the ipv6 resources, I modelled them exactly as per the current ipv4 resources.

Reply:
Never mind, found out the reason.
So, I can only see one possible solution: go back to the 3-variable version for ipv6 addr:
1. `var.enable_ipv6` as a global switch
2. `var.create_ipv6_addr`- to create the *new* address in the module
3. `var.ipv6_address` - for own predefined address.

Does this look ok? If yes, I'll push the required changes.

Reply:
Yeah I think that's the only option.

Reply:
I have pushed the changes with 3 variables.
Please have a look and let me know if it looks OK or needs further changes.